microsoft/vscode
# Feature: Shift Shift, Ctrl Ctrl, Alt Alt Keybinds (Double press modifer keybind support)

This PR fixes #5280 How to detect shift shift key in VSCode?

Proposed changes:

Implemented support for quick double presses of a modifier key without unit test regressions or noticeable performance hits, while following existing architecture as closely as possible.

This implementation improves UX experience for users migrating from JetBrains platform

Keybindings.json
```
    {
        "key": "shift shift",
        "command": "workbench.action.quickOpen"
    },
    {
        "key": "ctrl ctrl",
        "command": "workbench.action.showCommands"
    },
    {
        "key": "alt alt",
        "command": "workbench.action.quickOpen"
    },
```

- Implemented new keyboard event that checks for keyUp (critical for double press detection, second key MUST be released)
- Implemented additional keyup resolver for only modifiers, this will allow backwards compatibility with existing keybind system which uses keydown
- MacLinuxKeyboardMapper `getDispatchString()` had logic in places inconsistent with windowsKeyboardMapper, fixed

How to test:
- Run all unit tests, should all pass
- Test the above Keybindings.json
- Single modifier press should not activate chord (check notification bar bottom left no changes)
- Shift+Alphanumeric should work as usual
- Shift, Shift+Alphanumeric should work as usual no matter how quickly Shift, Shift is pressed (because of keyUp only detection)

Performance Test:
- Hold backspace on a large string; should be no lag
- Type fast; should be no lag

![image](https://user-images.githubusercontent.com/12622625/105950019-8e326c80-60d2-11eb-8c7c-2ecc24a8c03e.png)

I'm finding my self pressing "Shift Shift" in VSCode release when cross referencing code for this PR 😂😂😂🤦‍♂️
